### ISSUE TITLE
Rule proposal: no-promise-executor-return extension

### ISSUE THREAD

**Issue by PaperStrike** [typescript-eslint/typescript-eslint#4160]
### Description

There's a rule [`no-promise-executor-return`](https://eslint.org/docs/rules/no-promise-executor-return) in native ESLint, which disallows returning values from Promise executor functions.

This is **annoying** when you want to write a one line Promise that wraps a function, and you know the return value of the executor won't be used, while ESLint keeps complain about it. See the Pass section below for examples.

I think `@typescript-eslint/eslint-plugin` is the right place to extend this rule, as it can use the type information to know if the function returns `void`. When passing a value that is `void`, ESLint should know we ain't mistakenly expecting the value to be used.

Thank you!

### Fail

```ts
const returnNonVoid = () => 1;
new Promise(() => returnNonVoid());

new Promise(() => {
  return 1;
});
```

### Pass

```ts
const returnVoid = () => {};
new Promise(() => returnVoid());

new Promise(() => {
  return;
});
```

Safely use some one-liners:

```ts
new Promise((resolve) => queueMicrotask(resolve));
new Promise((resolve) => window.addEventListener('popstate', resolve));
new Promise((resolve) => void setTimeout(resolve));
```

**Comment by bradzacher** [comment]
Type information is *really* heavy, so implementing it in an extension rule has to provide some significant benefits, or else it's just not worth while.

From my understanding - you want to be able to use a body-less arrow function as the argument to a promise constructor. I don't think there's enough added value here for it to make sense for us to implement an entire extension rule.

I feel like this is generally safe enough that that it could just be an option in the base rule - and a quick search reveals that they would be somewhat open to an option to allow bodyless arrow functions that "return" a function call - https://github.com/eslint/eslint/issues/13668
The issue was closed as nobody implemented it, and the ESLint team closes out things that people don't look to implement quickly. But if you're so motivated, I'm sure they'd welcome and issue to confirm followed by a PR.

For that reason in addition to the above "low value" reason - I'm going to reject this proposal for this project.
